### PR TITLE
Add support for 'copy' method and default parameters

### LINKED ISSUE #21: Support for generic copy() method
I tried to compile Leon/Stainless code using the generic `copy()` method for generic case classes (see [here](http://stackoverflow.com/questions/7249396/how-to-clone-a-case-class-instance-and-change-just-one-field-in-scala)), but it seems that it is not yet in the Stainless library. Is it possible to provide support for this construct?

### PR DESCRIPTION
**Pull request by romac**
Close #21

### CODE REVIEW

**Comment by romac** [comment]
@samarion Ready for review.

**Comment by romac** [comment]
@samarion Ready for review.

File: frontends/dotty/src/main/scala/stainless/frontends/dotc/ASTExtractors.scala [hunk 1]
```diff
@@ -442,8 +456,10 @@ trait ASTExtractors {
             !dd.symbol.is(Label)
           ) || (
             (dd.symbol is Synthetic) &&
-            (dd.symbol is Implicit) &&
-            !(getAnnotations(tpt.symbol) contains "ignore")
+            (
+              canExtractSynthetic(dd.symbol) ||
+              !(getAnnotations(tpt.symbol) contains "ignore")
```
Comment (samarion) [inline frontends/dotty/src/main/scala/stainless/frontends/dotc/ASTExtractors.scala]: Why did this become an `or`?

File: core/src/main/scala/stainless/extraction/imperative/EffectsChecker.scala [hunk 2]
```diff
@@ -4,13 +4,36 @@ package stainless
...
 
-  def checkFunction(fd: FunDef)(symbols: Symbols, effects: EffectsAnalysis): Unit = {
+  protected def checkFunction(fd: FunDef)(symbols: Symbols, effects: EffectsAnalysis): CheckResult = {
     import symbols._
     import effects._
 
+    def isMutableSynthetic(id: Identifier): Boolean = {
+      val fd = symbols.functions(id)
+      (fd.flags contains Synthetic) && (fd.params match {
+        case receiver :: _ =>
```
Comment (samarion) [inline core/src/main/scala/stainless/extraction/imperative/EffectsChecker.scala]: Looking only at the receiver seems a bit brittle. Why not generalize to all params?

File: core/src/main/scala/stainless/extraction/imperative/AntiAliasing.scala [hunk 3]
```diff
@@ -382,7 +390,8 @@ trait AntiAliasing
       rec(effect.receiver, effect.target.path)
     }
 
-    transformer.transform(updateFunction(Outer(fd), Environment.empty).toFun)
+    val res = transformer.transform(updateFunction(Outer(fd), Environment.empty).toFun)
+    Some(res)
```
Comment (samarion) [inline core/src/main/scala/stainless/extraction/imperative/AntiAliasing.scala]: Remove superfluous let binding?

File: frontends/scalac/src/main/scala/stainless/frontends/scalac/CodeExtraction.scala [hunk 4]
```diff
@@ -319,18 +319,12 @@ trait CodeExtraction extends ASTExtractors {
       case EmptyTree =>
         // ignore
 
-      case t if (
-        (annotationsOf(t.symbol) contains xt.Ignore) ||
-        (t.symbol.isSynthetic && !t.symbol.isImplicit)
-      ) =>
+      case t if annotationsOf(t.symbol) contains xt.Ignore =>
```
Comment (samarion) [inline frontends/scalac/src/main/scala/stainless/frontends/scalac/CodeExtraction.scala]: Aren't you missing a condition here? What about `t.symbol.isSynthetic && !canExtractSynthetic(t.symbol)`?

File: frontends/dotty/src/main/scala/stainless/frontends/dotc/CodeExtraction.scala [hunk 5]
```diff
@@ -377,7 +373,8 @@ class CodeExtraction(inoxCtx: inox.Context, cache: SymbolsContext)(implicit val
     var flags = annotationsOf(sym) ++
       (if ((sym is Implicit) && (sym is Synthetic)) Set(xt.Inline, xt.Synthetic) else Set()) ++
       (if (sym is Inline) Set(xt.Inline) else Set()) ++
-      (if (!(sym is Method)) Set(xt.IsField(sym is Lazy)) else Set())
+      (if (!(sym is Method)) Set(xt.IsField(sym is Lazy)) else Set()) ++
+      (if (isCopyMethod(sym) || isDefaultGetter(sym)) Set(xt.Synthetic, xt.Inline) else Set())
```
Comment (samarion) [inline frontends/dotty/src/main/scala/stainless/frontends/dotc/CodeExtraction.scala]: I'm not sure the `copy` method should be inlined. It seems like that would be a bit surprising for the user, no? But the default getter should indeed be inlined.

File: frontends/scalac/src/main/scala/stainless/frontends/scalac/CodeExtraction.scala [hunk 6]
```diff
@@ -442,7 +436,8 @@ trait CodeExtraction extends ASTExtractors {
 
     var flags = annotationsOf(sym) ++
       (if (sym.isImplicit && sym.isSynthetic) Set(xt.Inline, xt.Synthetic) else Set()) ++
-      (if (sym.isAccessor) Set(xt.IsField(sym.isLazy)) else Set())
+      (if (sym.isAccessor) Set(xt.IsField(sym.isLazy)) else Set()) ++
+      (if (isDefaultGetter(sym) || isCopyMethod(sym)) Set(xt.Synthetic, xt.Inline) else Set())
```
Comment (samarion) [inline frontends/scalac/src/main/scala/stainless/frontends/scalac/CodeExtraction.scala]: Same remark as above about inlining the copy method.

**Comment by romac** [comment]
Looks like we should add a ADT's invariant as a precondition to the `copy` method, as otherwise such code goes through without errors:

```scala
  case class Foo(x: BigInt) {
    require(x != 0)
  }

  def updateFoo(foo: Foo) = {
    foo.copy(x = 0)
  }
```

**Comment by romac** [comment]
@samarion Rebased and ready for review/merge.

File: core/src/main/scala/stainless/extraction/inlining/FunctionInlining.scala [hunk 7]
```diff
@@ -64,7 +66,7 @@ trait FunctionInlining extends CachingPhase with IdentitySorts { self =>
             Let(vd, e, Assume(post, vd.toVariable).copiedFrom(fi)).copiedFrom(fi)
         }
 
-        val newBody = addPreconditionAssertion(addPostconditionAssumption(uncheckedBody))
+        val newBody = addPreconditionAssertion(addPostconditionAssumption(body))
```
Comment (samarion) [inline core/src/main/scala/stainless/extraction/inlining/FunctionInlining.scala]: We can't assume the post on `@synthetic` methods as it won't be checked anywhere. It should be inlined into an assertion here.

File: frontends/dotty/src/main/scala/stainless/frontends/dotc/CodeExtraction.scala [hunk 8]
```diff
@@ -377,7 +373,9 @@ class CodeExtraction(inoxCtx: inox.Context, cache: SymbolsContext)(implicit val
     var flags = annotationsOf(sym) ++
       (if ((sym is Implicit) && (sym is Synthetic)) Set(xt.Inline, xt.Synthetic) else Set()) ++
       (if (sym is Inline) Set(xt.Inline) else Set()) ++
-      (if (!(sym is Method)) Set(xt.IsField(sym is Lazy)) else Set())
+      (if (!(sym is Method)) Set(xt.IsField(sym is Lazy)) else Set()) ++
+      (if (isDefaultGetter(sym)) Set(xt.Synthetic, xt.Inline, xt.Annotation("specialMethod", Seq("defaultGetter"))) else Set()) ++
+      (if (isCopyMethod(sym)) Set(xt.Synthetic, xt.Inline, xt.Annotation("specialMethod", Seq("copy"))) else Set())
```
Comment (samarion) [inline frontends/dotty/src/main/scala/stainless/frontends/dotc/CodeExtraction.scala]: Let's not add the `specialMethod` flags as long as they're not necessary.

**Pull request merged**